src/Symfony/Component/Messenger/Middleware/DispatchAfterCurrentBusMiddleware.php

```diff
@@ -44,12 +44,13 @@ class DispatchAfterCurrentBusMiddleware implements MiddlewareInterface
     public function handle(Envelope $envelope, StackInterface $stack): Envelope
     {
         if (null !== $envelope->last(DispatchAfterCurrentBusStamp::class)) {
-            if (!$this->isRootDispatchCallRunning) {
-                throw new \LogicException(sprintf('You can only use a "%s" stamp in the context of a message handler.', DispatchAfterCurrentBusStamp::class));
+            if ($this->isRootDispatchCallRunning) {
+                $this->queue[] = new QueuedEnvelope($envelope, $stack);
+
+                return $envelope;
             }
-            $this->queue[] = new QueuedEnvelope($envelope, $stack);
 
-            return $envelope;
+            $envelope = $envelope->withoutAll(DispatchAfterCurrentBusStamp::class);
```

Comment:
Removing it from the envelope makes sense to me here for two reasons:
- even if it was explicitly configured with this stamp, it was not handled "transactionally". The userland bus might add it automatically by design and won't care about the context. So the stamp being present after the dispatch in the profiler might be confusing for the user.
- some logic could exist for messages dispatched transactionally with this stamp. Removing it in the case it was handled directly will prevent triggering such logic unexpectedly.